lib/common/src/Dom/Document.php

```diff
@@ -0,0 +1,1341 @@
+
+    // Regex patterns used for finding tags or extracting attribute values in an HTML string.
+    const HTML_FIND_TAG_WITHOUT_ATTRIBUTE_PATTERN = '/<%1$s[^>]*?>[^<]*(?:<\/%1$s>)?/i';
+    const HTML_FIND_TAG_WITH_ATTRIBUTE_PATTERN    = '/<%1$s [^>]*?\s*%2$s\s*=[^>]*?>[^<]*(?:<\/%1$s>)?/i';
+    const HTML_EXTRACT_ATTRIBUTE_VALUE_PATTERN    = '/%s=(?:([\'"])(?<full>.*)?\1|(?<partial>[^ \'";]+))/';
+    const HTML_FIND_TAG_DELIMITER                 = '/';
+
+    /**
+     * Pattern to match an Amp emoji together with its variant (amp4ads, amp4email, ...).
+     *
+     * @var string
+     */
+    const AMP_EMOJI_ATTRIBUTE_PATTERN = '/(<html [^>]*?)(' . Attribute::AMP_EMOJI_ALT . '|' . Attribute::AMP_EMOJI . ')([^\s^>]*)/iu';
+
+    // Attribute to use as a placeholder to move the emoji AMP symbol (⚡) over to DOM.
+    const EMOJI_AMP_ATTRIBUTE_PLACEHOLDER = 'emoji-amp';
+
+    /**
+     * Associative array of encoding mappings.
+     *
+     * Translates HTML charsets into encodings PHP can understand.
+     *
+     * @var string[]
+     */
+    const ENCODING_MAP = [
+        // Assume ISO-8859-1 for some charsets.
+        'latin-1' => 'ISO-8859-1',
+    ];
+
+    /**
+     * The original encoding of how the Amp\Dom\Document was created.
+     *
+     * This is stored to do an automatic conversion to UTF-8, which is
+     * a requirement for AMP.
+     *
+     * @var string
+     */
+    private $originalEncoding;
+
+    /**
+     * Store the placeholder comments that were generated to replace <noscript> elements.
+     *
+     * @see maybeReplaceNoscriptElements()
+     *
+     * @var string[]
+     */
+    private $noscriptPlaceholderComments = [];
+
+    /**
+     * Store whether mustache template tags were replaced and need to be restored.
+     *
+     * @see replaceMustacheTemplateTokens()
+     *
+     * @var bool
+     */
+    private $mustacheTagsReplaced = false;
+
+    /**
+     * Whether we had secured a doctype that needs restoring or not.
+     *
+     * This is an int as it receives the $count from the preg_replace().
+     *
+     * @var int
+     */
+    private $securedDoctype = 0;
+
+    /**
+     * Whether the self-closing tags were transformed and need to be restored.
+     *
+     * This avoids duplicating this effort (maybe corrupting the DOM) on multiple calls to saveHTML().
+     *
+     * @var bool
+     */
+    private $selfClosingTagsTransformed = false;
+
+    /**
+     * Store the emoji that was used to represent the AMP attribute.
+     *
+     * There are a few variations, so we want to keep track of this.
+     *
+     * @see https://github.com/ampproject/amphtml/issues/25990
+     *
+     * @var string
+     */
+    private $usedAmpEmoji;
+
+    /**
+     * Associative array of tweaks that will be run against the produced HTML output.
+     *
+     * This allows third-party code to fix DOMDocument problems that we don't yet account for.
+     *
+     * @var Tweak[]
+     */
+    private $tweaks;
```

Comment:
When a `CannotInlineRuntimeCss` error occurs, and SSR cannot be performed, I'm getting:

> `PHP Warning:  Invalid argument supplied for foreach() in /app/public/content/plugins/amp/vendor/amp/common/src/Dom/Document.php on line 467`

```suggestion
    private $tweaks = [];
```

Reply:
Nevertheless, I think the `tweaks` (or whatever we call them) should be postponed for now.

Reply:
Understood, I'll remove it, although I think we'd already have a handful of these tweaks in there right now.